python_modules/libraries/dagster-dg/dagster_dg/utils/filesystem.py

```diff
@@ -0,0 +1,110 @@
+import datetime
+import hashlib
+import time
+from collections.abc import Sequence
+from pathlib import Path
+from typing import Any, Callable, Optional
+
+from watchdog.events import FileSystemEvent, FileSystemEventHandler
+from watchdog.observers import Observer
+
+from dagster_dg.utils import (
+    DEFAULT_FILE_EXCLUDE_PATTERNS,
+    clear_screen,
+    hash_directory_metadata,
+    hash_file_metadata,
+)
+
+
+def hash_paths(
+    paths: Sequence[Path],
+    includes: Optional[Sequence[str]] = None,
+    excludes: Sequence[str] = DEFAULT_FILE_EXCLUDE_PATTERNS,
+) -> str:
+    """Hash the given paths, including their metadata.
+
+    Args:
+        paths: The paths to hash.
+        includes: A list of glob patterns to target, excluding files that don't match any of the patterns.
+        excludes: A list of glob patterns to exclude, including files that match any of the patterns. Defaults to
+            various Python-related files that are not relevant to the contents of the project.
+
+    Returns:
+        The hash of the paths.
+    """
+    hasher = hashlib.md5()
+    for path in paths:
+        if path.is_dir():
+            hash_directory_metadata(hasher, path, includes=includes, excludes=excludes)
+        else:
+            hash_file_metadata(hasher, path)
+    return hasher.hexdigest()
+
+
+class PathChangeHandler(FileSystemEventHandler):
+    """Basic handler that clears the screen and re-executes the callback when any of the watched paths change
+    in a way that would affect the hash of the paths. Passes the new hash to the callback.
```

Comment:
this is the part that I want to be sure i understand - what's an example of a change to a file that shouldn't trigger this? Would we get any benefits from bringing this into the hot reloading path that is used by dagster dev? And if we did, would there be a good way to do that given that it's in the core dagster package?

Reply:
>  what's an example of a change to a file that shouldn't trigger this?

Right now we hash on update time as well as contents, so I think all changes to non excluded files affect the hash - the main utility here is the include/exclude logic that we already use for hashing

Reply:
OK, got it, that makes sense. I will say that the comment here makes it sound more complicated than it is then - wasn't clear to me that 'affect the hash of the paths' just means 'a matching file changed'.

I also think watchdog gives you some decent ways to do include/exclude filtering as well, e.g. via a PatternMatchingEventHandler.

None of that is blocking though, just something to consider.